keps/sig-network/0031-20181017-kube-proxy-services-optional.md

```diff
@@ -0,0 +1,117 @@
+---
+kep-number: 0
+title: Make kube-proxy service abstraction optional
+authors:
+  - "@bradhoekstra"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: "@bradhoekstra"
+creation-date: 2018-10-17
+last-updated: 2018-10-17
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Make kube-proxy service abstraction optional
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Design](#design)
+      * [Testing](#testing)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+In a cluster that has a service mesh a lot of the work being done by kube-proxy is redundant and wasted.
+Specifically, services that are only reached via other services in the mesh will never use the service abstraction implemented by kube-proxy in iptables (or ipvs).
+By informing the kube-proxy of this, we can lighten the work it is doing and the burden on its proxy backend.
+
+## Motivation
+
+The motivation for the enhancement is to allow higher scalability in large clusters with lots of services that are making use of a service mesh.
+
+### Goals
+
+The goal is to reduce the load on:
+* The apiserver sending all services and endpoints to all kube-proxy pods
+* The kube-proxy having to deserialize and process all services and endpoints
+* The backend system (e.g. iptables) for whichever proxy mode kube-proxy is using
+
+### Non-Goals
+
+* Making sure the service is still routable via the service mesh
+* Preserving any kube-proxy functionality for any intentionally disabled Service, including but not limited to: externalIPs, external LB routing, nodePorts, externalTrafficPolicy, healthCheckNodePort, UDP, SCTP
```

Comment:
Do we have some supporting evidence, even anecdotal that this tradeoff is one that significant numbers of users are willing to make?  I don't want to filibuster this, but I want to know that we have real users for changes we make.

Reply:
I can see anyone offering services build on top of service meshes wanting this sort of feature, such as Istio as a service, or Knative, or any of the products that build on top Knative.

Knative as a specific example, or any framework that depends on a service mesh, may want to enable this as a default optimization of their deployment.

The idea for this came up in a talk with @mwielgus about rapidly scaling serverless workloads and he mentioned that updating large Endpoints objects in clusters with lots of nodes becomes a bottleneck.